Simulate BLE connection between central and peripheral zephyr samples

grifcj opened the pull request:
Changes to enable simulation of bluetooth connection between zephyr samples central_hr and periperhal_hr. I've been testing against main, [most recently this commit](https://github.com/zephyrproject-rtos/zephyr/commit/d6adbf9d0e0036b6d7f6e5760d439e864ed289af). Simulation should work without any modifications to project configs. I did most development with heart-rate samples, but also did a quick check with central_ht/peripheral_ht and those seemed to connect.

west commands to build zephyr images
```
west build -d ../build-central-hr -b nrf52840dk_nrf52840 -p always samples/bluetooth/central_hr
west build -d ../build-peripheral-hr -b nrf52840dk_nrf52840 -p always samples/bluetooth/peripheral_hr
```

Changes here are paired with updates to parent renode repository, which contains an example script for executing the demo. See [bluetooth-work branch in my fork](https://github.com/grifcj/renode/commits/bluetooth-work).

The changes primarily deal with updates for coordinating the radio with timers and PPI. The zephyr stack makes heavy use of PPI and rtc0, timer0, & timer1 for scheduling actions in bluetooth controller firmware. The firmware is fairly time aware, and expects radio actions to consume some amount of time and that simulation has enough resolution to adequately order events below 100 us. I found that a simulation quantum of 100 us would not function, but a 10 us one does. Perhaps there are ways around this, as it unfortunately affects overall simulation performance.

I've only tested with stock zephyr configuration of samples and with privacy and security disabled. Things that rely on additional hardware features beyond default config, e.g. TIFS hardware support, isn't included in these changes.

![image](https://user-images.githubusercontent.com/13724578/138049433-32a56709-51c2-485a-8a68-0f8ce7729c26.png)

![image](https://user-images.githubusercontent.com/13724578/138049098-8cc1ee63-ab24-4bf6-af88-5ba4535683de.png)

grifcj committed:
NRF52840_Radio: Use concurrent queue for thread-safety

grifcj committed:
NRF52840_Radio: Don't send address to packetptr. Use in-RAM representation.

grifcj committed:
NRF52840_Radio: Print received frame in noisy log

grifcj committed:
NRF52840_Radio: Label interrupt manager with RadioIrq

grifcj committed:
NRF52840_Radio: Add delays before triggering packet events on transmit

Code in interrupt handlers assume some time delay before packet goes
out, hence the resulting actions as triggered to PPI need some delay.
This change was found to ok for 10 us quantum.

grifcj committed:
NRF52840_Radio: Can only receive on frame per RxEnable

Remove while loop, just pull one frame off queue. Loop causes simulation
to deadlock because after receiving first frame, then radio is disabled
and the receive just enqueues again.

grifcj committed:
NRF52840_Radio: Add event delays to frame reception

Between RxEnable and PacketEnd and RadioDisable, embedded code may be
written in way that assumes there is some time passing. Have not
verified this like with transmit side, but seems a reasonable assumption
that it should be similar.

grifcj committed:
NRF52840_Radio: Set CRC Status to always good

grifcj committed:
NRF52840_Radio: Implement Channel property with BLE channel map

grifcj committed:
NRF52840_Radio: Handle difference between onAir and inRam packet

When sending to wireless medium, use in-air representation. When
reading/writing to memory use RAM representation.

Radio allows s1 field to be optionally included in on-air packet. This
becomes relevant for BLE data PDUs which may have 16 or 24 bit header.

grifcj commented:
Latest work here does a couple things.

It enables some simple soft device execution as described [here](https://github.com/renode/renode/issues/5)

Adding a basic ECB peripheral enables execution with CONFIG_BT_SMP. CONFIG_BT_PRIVACY also works. Work here being defined as heart-rate example between central/peripheral can connect and send notifications. This was against latest zephyr/main:d435340f1b.

mateusz-holenko commented:
@grifcj I merged fixes to `BasicClockSource` and `ComparingTimer` together with your (slightly refactored) tests upstream: 62fa1ef0110d5004f64cb01dae1c684b6426f25e.

mateusz-holenko commented:
> Cool, so it be ok if I rebase on latest master, drop the related commits from PR, and force push the PR branch? 

That would be great, thanks!

grifcj commented:
@mateusz-holenko As you can see, pushed changes up. Added a couple, the interrupt fix to RTC being the important one.

Spot checked against 3 sims and all appeared to work as intended / or at least equivalent to where I last left them. The last two configurations are less functional than the zephyr one. 

* Zephyr central-hr w/ peripheral-hr
  Connection functional and logging for HR notifications observed
* Nordic softdevice peripher-hr
  Wireshark shows tracing for advertising
* Zephyr central-hr w/ riot+nimble peripheral-hr
  Peripheral is advertising but not responding correctly to connect requests.

grifcj committed:
NRF52840_Radio: Extract radio event scheduling to method

Calculation of timing for transmit and receive side of radio events is
similar enough to share implementation.

grifcj committed:
NRF52840_Radio: Add CRCOk event on packet end

grifcj committed:
NRF52840_Radio: Add RSSI tasks and events. Fix to -10 dBm.

grifcj committed:
NRF52840_Radio: Add bit counter compare with single match

Bit counter can used for generating successive events. This
implementation just fires a single match event configured prior to
receiving radio frame.

grifcj committed:
NRF52840_Radio: Trigger Timer0 captures directly to avoid PPI delays

grifcj commented:
Latest commits enhance radio model and enable simulation of RIOT/NimBLE peripheral w/ Zephyr central. The existing Zephyr central / peripheral functionality remains. Nordic softdevice peripheral with zephyr central seemed more functional, but no functional connection.

Timing accuracy seems to require 1 us quantum, which slows down simulation considerably. 

![Screenshot from 2021-11-14 23-30-20](https://user-images.githubusercontent.com/13724578/141744734-1803326d-e3c7-419c-a6e1-2d093ed94970.png)
 
Couple of issues. 
* The direct timer0 captures are a hack, but it seems to be one way to get timer capture without delay. Capture must coincide with the end event and even a 1 us delay through next sync point is too much for NimBLE logic. An ISR fires and reads timer value immediately and expects it to be the captured packet end time.
* I'm using a nrf52840.repl that tags DEVICEADDR like so. This works for RIOT but not Zephyr. There seems to be an issue with word vs byte access. RIOT build was triggering byte access and I think Zephyr the opposite.
```
        Tag <0x100000a0, 0x100000a3> "DEVICEADDRTYPE" 0x1 # random address
        Tag <0x100000a4, 0x100000a4> "DEVICEADDR[0][0]" 0xDD
        Tag <0x100000a5, 0x100000a5> "DEVICEADDR[0][1]" 0xCC
        Tag <0x100000a6, 0x100000a6> "DEVICEADDR[0][2]" 0xBB
        Tag <0x100000a7, 0x100000a7> "DEVICEADDR[0][3]" 0xAA
```

mateusz-holenko commented:
I checked and merged upstream (with cosmetic formatting changes) commits fixing things in CLOCK, PPI and Timer.

I also took a look at the ECB model and it seems fine. I believe it should be fairly simple to add actual encryption logic after extracting it from [CC2538_Cryptoprocessor](https://github.com/renode/renode-infrastructure/blob/master/src/Emulator/Peripherals/Peripherals/Miscellaneous/CC2538_Cryptoprocessor.cs#L725) - I'll do that as a next step.

I'm not sure whether logger-related changes should be merged upstream. If you find them useful, could you extract them from this PR so that we can discuss them separately?

grifcj commented:
Rebased on latest master, dropping merged commits. Extracted logging to separate branch and will create ticket for further discussion / development.

mateusz-holenko commented:
I extended the ECB model with actual AES calculations and merged this together with RTC changes upstream.

What's left for review now are changes to the Radio controller model.

grifcj commented:
Rebased. Tested again zephyr and riot simulations. Everything still works. 

My platform file has some small mods to make RIOT simulation work. It needs the interrupt hookup for ECB peripheral. The DEVICEADDR stuff was described [above](https://github.com/renode/renode-infrastructure/pull/56#issuecomment-968637781)
``` diff
@@ -87,11 +87,15 @@ rng: Miscellaneous.NRF52840_RNG @ sysbus 0x4000D000
     -> nvic@13
 
 ecb: Miscellaneous.NRF52840_ECB @ sysbus 0x4000E000
+    -> nvic@14
 
 sysbus:
     init:
         ApplySVD @https://dl.antmicro.com/projects/renode/svd/NRF52840.svd.gz
         Tag <0x10000000, 0x10000FFF> "FICR"
         Tag <0x100000a0, 0x100000a3> "DEVICEADDRTYPE" 0x1 # random address
-        Tag <0x100000a4, 0x100000a7> "DEVICEADDR[0]" 0xAABBCCDD
+        Tag <0x100000a4, 0x100000a4> "DEVICEADDR[0][0]" 0xDD
+        Tag <0x100000a5, 0x100000a5> "DEVICEADDR[0][1]" 0xCC
+        Tag <0x100000a6, 0x100000a6> "DEVICEADDR[0][2]" 0xBB
+        Tag <0x100000a7, 0x100000a7> "DEVICEADDR[0][3]" 0xAA
         Tag <0x10001200, 0x10001203> "PSELRESET"

```

mateusz-holenko commented:
Most of the changes to `NRF52840_Radio` are now merged upstream - thanks again for your contribution @grifcj!

We added an automated test running the Zephyr peripheral/central HR sample.

For now we skipped the direct PPI events triggering, as those were quite hacky and seemed not to be necessary for the Zephyr demo. In the comment you mentioned that those were necessary to support RIOT stack correctly, is that right? Could you elaborate on this a bit more - perhaps we can come up with an alternative solution to the problem.

grifcj commented:
Some of the RIOT ISRs have code like this below, which rely on timer captures triggered by radio events. This particular one is in nimble/drivers/nrf52/src/ble_phy.c (I believe this was one of the problem ones, but didn't verify it again). Summary is that the simulated day from radio event to PPI task is 1 or 2 quantums, which allows too much ISR code to execute and results in incorrect simulated behavior.

![image](https://user-images.githubusercontent.com/13724578/151299738-3c09bd3e-4f94-465f-b93a-9b3c2e5963f1.png)

Just a bit more specificity on this particular scenario, the preprogrammed channels 26 and 27 tie radio events Address and End to captures on Timer0 CC1 and CC2. On hardware,  PPI propagation is synced to 16 MHz clock, so a max latency of 62.5 ns.  When run in simulation, the PPI model uses the "ExecuteInNearestSyncState" for executing the tasks triggered by events and hence the simulated latency will be dependent on chosen time quantum. If I recall correctly, it can be up to 2 quantum delay. So 2 us >> 62 ns, and the simulated CPU executes too much ISR code (which is fired by radio END event), passing the statements where the timer captures are evaluated. The direct writes to the timer tasks in the radio model bypass the sync delays, ensuring that that captures take place in time.

Zephyr on the other hand doesn't have any code that stresses the simulation models in this fashion, so it executes fine without these statements.
